cmd/krew/cmd/list.go

```diff
@@ -24,14 +24,94 @@ import (
 
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/kubectl/pkg/util/slice"
 
+	"sigs.k8s.io/krew/pkg/constants"
 	"sigs.k8s.io/krew/pkg/installation"
 )
 
+type Entry struct {
+	Name    string
+	Version string
+}
+
+func sortByName(e []Entry) []Entry {
+	sort.Slice(e, func(a, b int) bool {
+		return e[a].Name < e[b].Name
+	})
+	return e
+}
+
+// Consume produces a junk GroupVersionKin for obj.GetObjectKind().GroupVersionKind().Empty() check to eat in PrintObj()
+type Consume struct {
+	APIVersion string
+	Kind       string
+}
+
+func (c Consume) SetGroupVersionKind(kind schema.GroupVersionKind) {
+	c.APIVersion, c.Kind = kind.ToAPIVersionAndKind()
+}
+
+func (c Consume) GroupVersionKind() schema.GroupVersionKind {
+	return schema.FromAPIVersionAndKind(c.APIVersion, c.Kind)
+}
+
+func (e Entry) GetObjectKind() schema.ObjectKind {
+	return Consume{constants.CurrentAPIVersion, constants.PluginKind}
+}
+
+func (e Entry) DeepCopyObject() runtime.Object {
+	newObj := Entry{e.Name, e.Version}
+	return newObj
+}
+
+type ListFlags struct {
+	JSONYamlPrintFlags *genericclioptions.JSONYamlPrintFlags
+	OutputFormat       *string
+}
+
+func NewListFlags() *ListFlags {
+	outputFormat := ""
+	return &ListFlags{
+		JSONYamlPrintFlags: genericclioptions.NewJSONYamlPrintFlags(),
+		OutputFormat:       &outputFormat,
+	}
+}
+
+func (f *ListFlags) AllowedFormats() []string {
+	formats := f.JSONYamlPrintFlags.AllowedFormats()
+	return formats
+}
+
+func (f *ListFlags) AddFlags(c *cobra.Command) {}
+
+func (f *ListFlags) ToPrinter() (printers.ResourcePrinter, error) {
+	outputFormat := *f.OutputFormat
+	var printer printers.ResourcePrinter
+
+	switch outputFormat {
+	case "json":
+		printer = &printers.JSONPrinter{}
+	case "yaml":
+		printer = &printers.YAMLPrinter{}
```

Comment:
with these added, we'll need additional integration tests next to the existing ones to verify the flags works as intended (and returns valid correct json and yaml that can be parsed into an object, say, like an `interface{}{}`).

this would make sure we're not printing a table for yaml/json cases, similarly we can see how these handle "empty" outputs as part of the tests.

Reply:
I'll draft up some tests.